### PR TITLE
Add support for "functions" usage in ChatCompletion

### PR DESCRIPTION
**Pull request by itaiMona**
Note that the main change here isn't about supporting "functions" logic more than just not crushing on their usage. 

This is since there are no specific metrics to be calculated right now about function usage, and all new fields are created by field build functions on the server side.

@MichalBS-mona and @HungryWizard this is a good opportunity for you guys to learn some more about this package as well.

### CODE REVIEW

File: tests/test_completion.py [hunk 1]
```diff
@@ -329,7 +332,12 @@ def test_export_prompt():
         _DEFAULT_CONTEXT_CLASS,
         {"export_prompt": True},
         mona_clients_getter=get_mock_mona_clients_getter(
-            (_get_mona_message(input=_remove_irrelevant_input_keys(_DEFAULT_INPUT)),), ()
+            (
+                _get_mona_message(
+                    input=_remove_irrelevant_input_keys(_DEFAULT_INPUT)
+                ),
+            ),
+            (),
```
Comment (HungryWizard) [inline tests/test_completion.py:345]: I see it's only format, but I would use kwargs so it'll be clear what does "()" stands for

File: tests/test_chat_completion.py [hunk 2]
```diff
@@ -335,7 +337,12 @@ def test_export_prompt():
         _DEFAULT_CONTEXT_CLASS,
         {"export_prompt": True},
         mona_clients_getter=get_mock_mona_clients_getter(
-            (_get_mona_message(input=_remove_irrelevant_input_keys(_DEFAULT_INPUT)),), ()
+            (
+                _get_mona_message(
+                    input=_remove_irrelevant_input_keys(_DEFAULT_INPUT)
+                ),
+            ),
+            (),
```
Comment (HungryWizard) [inline tests/test_chat_completion.py:345]: Across the board

File: mona_openai/util/oop_util.py
Comment (HungryWizard) [inline mona_openai/util/oop_util.py:1]: Why we don't need this anymore?

File: mona_openai/endpoints/completion.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 from ..analysis.privacy import get_privacy_analyzers, PrivacyAnalyzer
 from ..analysis.profanity import get_has_profanity, get_profanity_prob
 from ..analysis.textual import get_textual_analyzers, TextualAnalyzer
-from ..util.oop_util import create_combined_object
+from ..analysis.util import create_combined_analyzer
```
Comment (HungryWizard) [inline mona_openai/endpoints/completion.py:11]: We don't use this convention in other parts of our code
Is that on purpose?

File: mona_openai/analysis/profanity.py [hunk 4]
```diff
@@ -2,14 +2,25 @@
...
 """
 from collections.abc import Iterable
+from typing import Optional
 from profanity_check import predict_prob, predict
 
 _DECIMAL_PLACES = 2
 
 
-def get_profanity_prob(texts: Iterable[str]) -> tuple[float, ...]:
+def clear_nones(texts):
```
Comment (HungryWizard) [inline mona_openai/analysis/profanity.py]: Maybe move to a general util file

File: examples/chat_completion/chat_completion_functions.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+openai.api_key = environ.get("OPEN_AI_KEY")
+
+MONA_API_KEY = environ.get("MONA_API_KEY")
+MONA_SECRET = environ.get("MONA_SECRET")
+MONA_CREDS = {
+    "key": MONA_API_KEY,
+    "secret": MONA_SECRET,
+}
+
+# This is the name of the monitoring class on Mona
```
Comment (HungryWizard) [inline examples/chat_completion/chat_completion_functions.py]: Period

File: examples/chat_completion/chat_completion_functions.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        )
+        second_response = monitored_chat_completion.create(
+            model="gpt-3.5-turbo-0613",
+            messages=messages,
+        )
+        return second_response
+
+
+print(run_conversation())
```
Comment (HungryWizard) [inline examples/chat_completion/chat_completion_functions.py:101]: Why is this needed?

File: mona_openai/analysis/util.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                    if inspect.ismethod(func) or inspect.isfunction(func):
+                        results.append(func(*args, **kwargs))
+                return tuple(results)
+
+            return method
+
+    return CombinedObject(instances)
+
+
+def getAnalyzers(
```
Comment (HungryWizard) [inline mona_openai/analysis/util.py]: Why camelCase?

File: mona_openai/analysis/util.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
+from typing import Optional
+from collections.abc import Iterable
+import inspect
```
Comment (HungryWizard) [inline mona_openai/analysis/util.py:1]: I guess isort wouldn't allow this
Why not connect this repo to a pipeline similar to what we have on Github?

File: examples/chat_completion/chat_completion_functions.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
+from os import environ
+import openai
+import json
+
+from mona_openai import monitor
```
Comment (HungryWizard) [inline examples/chat_completion/chat_completion_functions.py:6]: Won't having the same name for the package and a file within, might cause confusion?

File: examples/chat_completion/chat_completion_functions.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+MONITORING_CONTEXT_NAME = "MONITORED_CHAT_COMPLETION_USE_CASE_NAME"
+
+monitored_chat_completion = monitor(
+    openai.ChatCompletion,
+    MONA_CREDS,
+    MONITORING_CONTEXT_NAME,
+)
+
+
+def get_current_weather(location, unit="fahrenheit"):
```
Comment (HungryWizard) [inline examples/chat_completion/chat_completion_functions.py:27]: Are those are like examples for possible usages of our package or am I missing something?

File: examples/chat_completion/chat_completion_functions_rest.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+# Set up the API endpoint URL and authentication headers
+url = "https://api.openai.com/v1/chat/completions"
+headers = {
+    "Content-Type": "application/json",
+    "Authorization": f"Bearer {environ.get('OPEN_AI_KEY')}",
+}
+
+
+def get_current_weather(location, unit="fahrenheit"):
```
Comment (HungryWizard) [inline examples/chat_completion/chat_completion_functions_rest.py:32]: Second time I'm seeing this code

File: examples/chat_completion/chat_completion_functions_rest.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+    # the response or the exception. When we later do that, the logger will
+    # actually calculate all the relevant metrics and will send them to
+    # Mona.
+    response_logger, exception_logger = mona_logger.log_request(
+        data, additional_data={"customer_id": "A531251"}
+    )
+
+    try:
+        # Send the request to the API
+        response = requests.post(url, headers=headers, json=data)
```
Comment (HungryWizard) [inline examples/chat_completion/chat_completion_functions_rest.py:89]: Why not with the OpenAI libary?

File: examples/chat_completion/chat_completion_functions_rest.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+        # Check for HTTP errors
+        response.raise_for_status()
+
+        # Log response to Mona
+        response_logger(response.json())
+        print(response.json()["choices"][0]["message"])
+
+    except Exception:
+        # Log exception to Mona
```
Comment (HungryWizard) [inline examples/chat_completion/chat_completion_functions_rest.py:99]: Periods, across the board

File: examples/chat_completion/chat_completion_functions_rest.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+            response_logger(second_response.json())
+            print(second_response.json()["choices"][0]["message"])
+            return second_response.json()
+
+        except Exception:
+            # Log exception to Mona
+            exception_logger()
+
+
+print(run_conversation())
```
Comment (HungryWizard) [inline examples/chat_completion/chat_completion_functions_rest.py:161]: Ditto

File: examples/chat_completion/chat_completion_functions.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+                    "unit": {
+                        "type": "string",
+                        "enum": ["celsius", "fahrenheit"],
+                    },
+                },
+                "required": ["location"],
+            },
+        }
+    ]
+    response = monitored_chat_completion.create(
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions.py:64]: Add empty line above

File: examples/chat_completion/chat_completion_functions.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+
+def get_current_weather(location, unit="fahrenheit"):
+    """Get the current weather in a given location"""
+    weather_info = {
+        "location": location,
+        "temperature": "72",
+        "unit": unit,
+        "forecast": ["sunny", "windy"],
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions.py:33]: nit - I find it weird that the temperature and forecast are hardcoded in the example

File: examples/chat_completion/chat_completion_functions.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+            },
+        }
+    ]
+    response = monitored_chat_completion.create(
+        model="gpt-3.5-turbo-0613",
+        messages=messages,
+        functions=functions,
+        function_call="auto",
+    )
+    response_message = response["choices"][0]["message"]
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions.py:70]: Can we extract and use the _get_only_choice util to access the message? Same goes for any example of this client's usage were we access the first choice, I feel like it's repeating itself

File: examples/chat_completion/chat_completion_functions.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+        "forecast": ["sunny", "windy"],
+    }
+    return json.dumps(weather_info)
+
+
+def run_conversation():
+    messages = [
+        {"role": "user", "content": "What's the weather like in Boston?"}
+    ]
+    functions = [
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions.py:42]: I think messages and functions should be defined outside run_conversion

File: examples/chat_completion/chat_completion_functions.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    ]
+    response = monitored_chat_completion.create(
+        model="gpt-3.5-turbo-0613",
+        messages=messages,
+        functions=functions,
+        function_call="auto",
+    )
+    response_message = response["choices"][0]["message"]
+
+    if response_message.get("function_call"):
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions.py:72]: nit - keep response_message.get("function_call") in a variable and reuse in lines 74 and 77

File: examples/chat_completion/chat_completion_functions.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+        available_functions = {
+            "get_current_weather": get_current_weather,
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions.py:74]: I feel like this should be defined under functions, this is some kind of two sources of truth - I can't think of a case where the functions we pass when creating monitored chat completion are different from the available functions... 
I'd add another "function" key in each function dict with the actual function, create the functions param as it is expected, and in line 75 get the function to call from that list.
WDYT? LMK if unclear

File: examples/chat_completion/chat_completion_functions.py [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        function_name = response_message["function_call"]["name"]
+        fuction_to_call = available_functions[function_name]
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions.py:77]: Is there only one function call per response (if any)?

File: examples/chat_completion/chat_completion_functions.py [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+        function_args = json.loads(
+            response_message["function_call"]["arguments"]
+        )
+        function_response = fuction_to_call(
+            location=function_args.get("location"),
+            unit=function_args.get("unit"),
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions.py:83]: Those two lines are specific for get_current_weather, which although is the only available function in this example, I think should be generalized. Maybe use **function_args?

File: examples/chat_completion/chat_completion_functions.py (hunk 19 above)
Comment (MichalBS-mona) [reply examples/chat_completion/chat_completion_functions.py:72]: What do we do if there isn't a function call in the response?

File: examples/chat_completion/chat_completion_functions.py [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+        )
+
+        messages.append(response_message)
+        messages.append(
+            {
+                "role": "function",
+                "name": function_name,
+                "content": function_response,
+            }
+        )
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions.py:93]: Why are we adding both the complete response and this function dict to messages?

File: examples/chat_completion/chat_completion_functions.py (hunk 6 above)
Comment (MichalBS-mona) [reply examples/chat_completion/chat_completion_functions.py:101]: +1 - why printing?

File: examples/chat_completion/chat_completion_functions_rest.py [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+        "model": "gpt-3.5-turbo-0613",
+        "n": 1,
+    }
+
+    # The log_request function returns two other function for later logging
+    # the response or the exception. When we later do that, the logger will
+    # actually calculate all the relevant metrics and will send them to
+    # Mona.
+    response_logger, exception_logger = mona_logger.log_request(
+        data, additional_data={"customer_id": "A531251"}
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions_rest.py:84]: What's this customer id? Do we want to keep it outside the function maybe to be filled in by the user?

File: examples/chat_completion/chat_completion_functions_rest.py [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+# Get Mona logger
+mona_logger = get_rest_monitor(
+    "ChatCompletion",
+    MONA_CREDS,
+    MONITORING_CONTEXT_NAME,
+)
+
+# Set up the API endpoint URL and authentication headers
+url = "https://api.openai.com/v1/chat/completions"
+headers = {
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions_rest.py:26]: url and headers can (should?) be all caps

also i'd give a more indicative name than "url" - would make the usage more readable

File: examples/chat_completion/chat_completion_functions_rest.py [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+    try:
+        # Send the request to the API
+        response = requests.post(url, headers=headers, json=data)
+
+        # Check for HTTP errors
+        response.raise_for_status()
+
+        # Log response to Mona
+        response_logger(response.json())
+        print(response.json()["choices"][0]["message"])
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions_rest.py:96]: Maybe init a "response_message" var outside the try-catch block with None value and assign it here, your'e repeating this  in line 101

File: examples/chat_completion/chat_completion_functions_rest.py [hunk 27]
```diff
@@ -0,0 +1,160 @@
...
+                        "enum": ["celsius", "fahrenheit"],
+                    },
+                },
+                "required": ["location"],
+            },
+        }
+    ]
+
+    # Set up the request data
+    data = {
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions_rest.py:70]: Let's rename request_data

File: examples/chat_completion/chat_completion_functions_rest.py [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+        messages.append(response_message)
+        messages.append(
+            {
+                "role": "function",
+                "name": function_name,
+                "content": function_response,
+            }
+        )
+
+        data = {
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions_rest.py:127]: Isn't this identical to data defined in line 69? Why are we doing this again?

File: examples/chat_completion/chat_completion_functions_rest.py [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+            second_response.raise_for_status()
+
+            # Log response to Mona
+            response_logger(second_response.json())
+            print(second_response.json()["choices"][0]["message"])
+            return second_response.json()
+
+        except Exception:
+            # Log exception to Mona
+            exception_logger()
```
Comment (MichalBS-mona) [inline examples/chat_completion/chat_completion_functions_rest.py:158]: Lines 69-99 and 126-157 are very much the same - can be extracted to another function.

File: mona_openai/analysis/util.py [hunk 30]
```diff
@@ -0,0 +1,98 @@
...
+                return none_func
+
+            attrs[attr_name] = make_none_func(attr_value)
+
+    # Add is_none_analyzer method that always returns True, after the loop to
+    # make sure it overrides any such original method.
+    def is_none_analyzer(self):
+        return True
+
+    attrs["is_none_analyzer"] = is_none_analyzer
```
Comment (MichalBS-mona) [inline mona_openai/analysis/util.py]: Why not add this already when defining attrs in line 23?

File: mona_openai/analysis/util.py [hunk 31]
```diff
@@ -0,0 +1,98 @@
+from typing import Optional
+from collections.abc import Iterable
+import inspect
+from .analyzer import Analyzer
+from functools import wraps
+
+
+def _get_none_analyzer_class(base_class, init_func):
```
Comment (MichalBS-mona) [inline mona_openai/analysis/util.py]: Very cool

File: mona_openai/analysis/util.py [hunk 32]
```diff
@@ -0,0 +1,98 @@
...
+    messages are the same.
+
+    TODO(itai): In the future we might want to introduce specific analyses for
+        function calls, which can be done within existing analyzers or with
+        new analyzers.
+    """
+    attrs = {"__init__": init_func}
+
+    for attr_name, attr_value in base_class.__dict__.items():
+        if callable(attr_value) and not attr_name.startswith("_"):
```
Comment (MichalBS-mona) [inline mona_openai/analysis/util.py:36]: Do we not care about non-callable attributes?

File: mona_openai/analysis/util.py [hunk 33]
```diff
@@ -0,0 +1,98 @@
...
+    the message describes a function call.
+
+    In short, these None analyzers allow us to keep the code flow as if all
+    messages are the same.
+
+    TODO(itai): In the future we might want to introduce specific analyses for
+        function calls, which can be done within existing analyzers or with
+        new analyzers.
+    """
+    attrs = {"__init__": init_func}
```
Comment (MichalBS-mona) [inline mona_openai/analysis/util.py]: Why not keep the none init function inside the base class as _none_init? I think that it should be an attribute of any base class that's passed in here, no?

File: mona_openai/endpoints/chat_completion.py [hunk 34]
```diff
@@ -164,6 +173,7 @@ def _get_full_textual_analysis(
                 if analyzer.get_word_count() > 0
                 else 0.0
                 for analyzer in answers_textual_analyzers
+                if not analyzer.is_none_analyzer()
```
Comment (MichalBS-mona) [inline mona_openai/endpoints/chat_completion.py:176]: Can we iterate combined_messages instead of answers_textual_analyzers and save this line? Or am I understanding create_combined_analyzer wrong?

**Comment by MichalBS-mona** [comment]
@itaiMona Are you planning to update the docs with the new usages?

**Comment by itaiMona** [comment]
@MichalBS-mona what usage is that? Can you explain what you would expect to see in the docs?

File: tests/test_completion.py (hunk 1 above)
Comment (itaiMona) [reply tests/test_completion.py:345]: This isn't really relevant for this PR. Leaving TODO

File: examples/chat_completion/chat_completion_functions.py (hunk 9 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions.py:6]: I don't see a problem with it... Also - not really relevant to this PR

File: examples/chat_completion/chat_completion_functions.py (hunk 10 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions.py:27]: Yes these files are example code files for users to have simpler boilerplate options.

File: examples/chat_completion/chat_completion_functions.py (hunk 16 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions.py:33]: I copy pasted this example from OpenAI's official docs about "functions". I think keeping the same example as in their docs will make things simpler and more familiar for the reader

File: examples/chat_completion/chat_completion_functions.py (hunk 18 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions.py:42]: See above - I used the example code as given by OpenAI's docs

File: examples/chat_completion/chat_completion_functions.py (hunk 17 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions.py:70]: This is an example code for the user to use for boilerplate, it isn't part of the package. This means we will need to expose that helper function as part of what the package exposes... This package doesn't exist to provide users with functional utilities... and exposing such a function sounds very confusing to me... So I'm against that.

Also see my comments regarding this being OpenAI's example

File: examples/chat_completion/chat_completion_functions.py (hunk 19 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions.py:72]: This is example usage code. I'm OK with this code failing due to something like that. It has nothing to do with the use of our package if that's the case. We're not trying to provide the user here with best in class example code, just with some way to easily start using our package and understand its ins and outs.

File: examples/chat_completion/chat_completion_functions.py (hunk 20 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions.py:74]: See all above comments.

File: examples/chat_completion/chat_completion_functions.py (hunk 21 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions.py:77]: yes. A message can be a function call or not. It cannot be multiple function calls

File: examples/chat_completion/chat_completion_functions.py (hunk 22 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions.py:83]: Again, unrelated to what we're trying to achieve in this file.

File: examples/chat_completion/chat_completion_functions.py (hunk 23 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions.py:93]: This is how you use Chat Completion - you keep appending the messages to a list so that whenever you call the API, you provide the "history" of the current conversation.

File: examples/chat_completion/chat_completion_functions.py (hunk 6 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions.py:101]: This is just something we do in all examples. Print the result. Really doesn't have any meaning to it.

File: examples/chat_completion/chat_completion_functions_rest.py (hunk 25 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions_rest.py:26]: Keeping this like every other example we have... 

Sorry for not explaining before that these files are just simple script examples. They shouldn't adhere to the same style guide as the actual code in the package

File: examples/chat_completion/chat_completion_functions_rest.py (hunk 27 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions_rest.py:70]: Let's leave this file alone 😅

File: examples/chat_completion/chat_completion_functions_rest.py (hunk 24 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions_rest.py:84]: Same as every other example file...

File: examples/chat_completion/chat_completion_functions_rest.py (hunk 12 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions_rest.py:89]: This is a "rest" example, as the file name suggests.

**Comment by MichalBS-mona** [comment]
@itaiMona there are new params we can pass to monitored_chat_completion.create ( functions, function_call), isn't it worth mentioning in the docs?

File: examples/chat_completion/chat_completion_functions_rest.py (hunk 26 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions_rest.py:96]: See other answers...

File: examples/chat_completion/chat_completion_functions_rest.py (hunk 13 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions_rest.py:99]: Leaving this... There's really no need for such styling changes

File: examples/chat_completion/chat_completion_functions_rest.py (hunk 28 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions_rest.py:127]: It's just a way to test again a usage that returns the function answer... I can see readers who might be interested in this specific part and will want to just copy that.

File: examples/chat_completion/chat_completion_functions_rest.py (hunk 29 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions_rest.py:158]: They can, but that's not what we're about in these files.

File: examples/chat_completion/chat_completion_functions_rest.py (hunk 14 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions_rest.py:161]: Ditto

File: mona_openai/analysis/profanity.py (hunk 4 above)
Comment (itaiMona) [reply mona_openai/analysis/profanity.py]: I did change to private - no need to expose this. But also no general use of this so I don't see a good reason to put in util. It's too basic.

File: mona_openai/analysis/util.py (hunk 8 above)
Comment (itaiMona) [reply mona_openai/analysis/util.py:1]: Thanks! I forgot to run isort. Fixed.

File: mona_openai/analysis/util.py (hunk 31 above)
Comment (itaiMona) [reply mona_openai/analysis/util.py]: Thank you

File: mona_openai/analysis/util.py (hunk 32 above)
Comment (itaiMona) [reply mona_openai/analysis/util.py:36]: No. At least not right now...

File: mona_openai/analysis/util.py (hunk 30 above)
Comment (itaiMona) [reply mona_openai/analysis/util.py]: Good idea

File: mona_openai/analysis/util.py (hunk 33 above)
Comment (itaiMona) [reply mona_openai/analysis/util.py]: Cool idea!

File: mona_openai/analysis/util.py (hunk 7 above)
Comment (itaiMona) [reply mona_openai/analysis/util.py]: Good catch!

File: mona_openai/endpoints/chat_completion.py (hunk 34 above)
Comment (itaiMona) [reply mona_openai/endpoints/chat_completion.py:176]: I can't see how this will help us.

File: mona_openai/endpoints/completion.py (hunk 3 above)
Comment (itaiMona) [reply mona_openai/endpoints/completion.py:11]: What convention?

File: mona_openai/util/oop_util.py
Comment (itaiMona) [reply mona_openai/util/oop_util.py:1]: The logic that was here isn't really about general OOP anymore, it's really specific to analyzers.

**Review by itaiMona — COMMENTED** [review]
@HungryWizard @MichalBS-mona Thanks for all the comments. I think there's one question that's still open. Looking forward to your approvals :)

**Comment by itaiMona** [comment]
@MichalBS-mona there are no new parameters that can be passed to our package - just to OpenAI's SDK. We are still just wrapping that SDK.

**Comment by MichalBS-mona** [comment]
LGTM

File: mona_openai/endpoints/completion.py (hunk 3 above)
Comment (HungryWizard) [reply mona_openai/endpoints/completion.py:11]: Using ".." to go up the tree from the current file when importing, instead of giving the "full" name from the root
 (from mona_openai.util.oop_util import create_combined_objectc"

Comment (itaiMona) [reply mona_openai/endpoints/completion.py:11]: Well I guess this is a discussion worth having, but again - not in this PR.

File: examples/chat_completion/chat_completion_functions_rest.py (hunk 11 above)
Comment (itaiMona) [reply examples/chat_completion/chat_completion_functions_rest.py:32]: Yes. This is example boilerplate code so expected.
